contracts/tec/contracts/src/MixinTECApprovalVerifier.sol

```diff
@@ -0,0 +1,200 @@
+    LibZeroExTransaction,
+    MSignatureValidator,
+    MTECApprovalVerifier
+{
+    using LibBytes for bytes;
+    using LibAddressArray for address[];
+
+    /// @dev Validates that the 0x transaction has been approved by all of the feeRecipients
+    ///      that correspond to each order in the transaction's Exchange calldata.
+    /// @param transaction 0x transaction containing salt, signerAddress, and data.
+    /// @param transactionSignature Proof that the transaction has been signed by the signer.
+    /// @param approvalExpirationTimeSeconds Array of expiration times in seconds for which each corresponding approval signature expires.
+    /// @param approvalSignatures Array of signatures that correspond to the feeRecipients of each order in the transaction's Exchange calldata.
+    function assertValidTECApprovals(
+        LibZeroExTransaction.ZeroExTransaction memory transaction,
+        bytes memory transactionSignature,
+        uint256[] memory approvalExpirationTimeSeconds,
+        bytes[] memory approvalSignatures
+    )
+        public
+        view
+    {
+        // Get the orders from the the Exchange calldata in the 0x transaction
+        LibOrder.Order[] memory orders = decodeFillDataOrders(transaction.data);
+
+        // No approval is required for non-fill methods
+        if (orders.length > 0) {
+            // Revert if approval is invalid for transaction orders
+            assertValidTransactionOrdersApproval(
+                transaction,
+                orders,
+                transactionSignature,
+                approvalExpirationTimeSeconds,
+                approvalSignatures
+            );
+        }
+    }
+
+    /// @dev Validates that the feeRecipients of a batch of order have approved a 0x transaction.
+    /// @param transaction 0x transaction containing salt, signerAddress, and data.
+    /// @param orders Array of order structs containing order specifications.
+    /// @param transactionSignature Proof that the transaction has been signed by the signer.
+    /// @param approvalExpirationTimeSeconds Array of expiration times in seconds for which each corresponding approval signature expires.
+    /// @param approvalSignatures Array of signatures that correspond to the feeRecipients of each order.
+    function assertValidTransactionOrdersApproval(
+        LibZeroExTransaction.ZeroExTransaction memory transaction,
+        LibOrder.Order[] memory orders,
+        bytes memory transactionSignature,
+        uint256[] memory approvalExpirationTimeSeconds,
+        bytes[] memory approvalSignatures
+    )
+        public
+        view
+    {
+        // Hash 0x transaction
+        bytes32 transactionHash = getTransactionHash(transaction);
+
+        // Create empty list of approval signers
+        address[] memory approvalSignerAddresses = new address[](0);
+
+        uint256 signaturesLength = approvalSignatures.length;
+        for (uint256 i = 0; i < signaturesLength; i++) {
```

Comment:
Should we assert that `approvalSignatures.length == approvalExpirationTimeSeconds.length` before iterating in lockstep? Should error out anyway, but I think it'll throw and waste the gas + won't get a helpful error message.